### PR TITLE
ParadeDB Support

### PR DESCRIPTION
**Pull request by itay-grudev**
This adds support for running a ParadeDB cluster using the CloudNativePG operator.

Closes: #2
Closes: #4
Closes: #5
Closes: #9

### CODE REVIEW

File: charts/cluster/templates/_bootstrap.tpl [hunk 1]
```diff
@@ -15,11 +15,27 @@ bootstrap:
...
-          {{- range .postInitApplicationSQL }}
-            {{- printf "- %s" . | nindent 6 }}
-          {{- end -}}
+        {{- range .postInitApplicationSQL }}
+          {{- printf "- %s" . | nindent 6 }}
+        {{- end -}}
+      {{- end }}
+    postInitTemplateSQL:
+      {{- if eq .Values.type "paradedb" }}
+      - ALTER DATABASE template1 SET search_path TO public,paradedb;
```
Comment (philippemnoel) [inline charts/cluster/templates/_bootstrap.tpl:33]: The extensions should be installed here too, no?

File: charts/cluster/test/paradedb-minio-backup-restore/00-minio_cleanup-assert.yaml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+apiVersion: batch/v1
```
Comment (philippemnoel) [inline charts/cluster/test/paradedb-minio-backup-restore/00-minio_cleanup-assert.yaml:1]: Is this file ParadeDB-related or just needs to be merged upstream?

File: charts/cluster/test/paradedb-minio-backup-restore/03-paradedb_test.yaml [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+        command: ['sh', '-c']
+        args:
+          - |
+            apk --no-cache add postgresql-client
+            RESULT=$(psql "$DB_URI" -t) <<-EOSQL
+              SELECT description
+              FROM search_idx.search('description:"bluetooth speaker"~1');
+            EOSQL
+            echo -$RESULT-
+            test "$RESULT" = " Bluetooth-enabled speaker"
```
Comment (philippemnoel) [inline charts/cluster/test/paradedb-minio-backup-restore/03-paradedb_test.yaml:27]: newline

File: charts/cluster/test/paradedb-minio-backup-restore/02-paradedb_write.yaml [hunk 4]
```diff
@@ -0,0 +1,35 @@
+apiVersion: batch/v1
```
Comment (philippemnoel) [inline charts/cluster/test/paradedb-minio-backup-restore/02-paradedb_write.yaml:1]: I wouldn't call this `_write`. This isn't ingesting data but just creating an index over an existing table

File: charts/cluster/test/paradedb-minio-backup-restore/03-paradedb_test.yaml [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+        image: alpine:3.19
+        command: ['sh', '-c']
+        args:
+          - |
+            apk --no-cache add postgresql-client
+            RESULT=$(psql "$DB_URI" -t) <<-EOSQL
+              SELECT description
+              FROM search_idx.search('description:"bluetooth speaker"~1');
+            EOSQL
+            echo -$RESULT-
```
Comment (philippemnoel) [inline charts/cluster/test/paradedb-minio-backup-restore/03-paradedb_test.yaml:26]: This is sweet. Could we put the whole ParadeDB Quickstart in the test? This way we're sure every feature is supported?

File: charts/cluster/test/postgresql-minio-backup-restore/05-recovery_backup_cluster.yaml [hunk 6]
```diff
@@ -19,7 +19,7 @@ recovery:
     encryption: ""
   s3:
     bucket: "mybucket"
-    path: "/v1"
+    path: "/postgresql-minio-backup-restore/v1"
```
Comment (philippemnoel) [inline charts/cluster/test/postgresql-minio-backup-restore/05-recovery_backup_cluster.yaml:22]: Is it correct that these files aren't ParadeDB-related but just need to be merged upstream?

File: charts/cluster/values.yaml [hunk 7]
```diff
@@ -17,6 +18,8 @@ version:
   timescaledb: "2.15"
   # -- If using PostGIS, specify the version
   postgis: "3.4"
+  # -- If using ParadeDB, specify the version
+  paradedb: "0.9.3"
```
Comment (philippemnoel) [inline charts/cluster/values.yaml:22]: In our own charts, is it possible to not hardcode this? We have an automated deployment process via CI in the `paradedb/helm-charts` repository which would be great to map here, for easy maintenance

**Comment by philippemnoel** [comment]
This looks promising! Thank you for making this. Just want to make sure to flag the issue list here: https://github.com/paradedb/helm-charts/issues

Could you please update the PR body with which issues are fixed by this PR, so we can track what's left for a subsequent PR?

File: charts/cluster/test/postgresql-minio-backup-restore/05-recovery_backup_cluster.yaml (hunk 6 above)
Comment (itay-grudev) [reply charts/cluster/test/postgresql-minio-backup-restore/05-recovery_backup_cluster.yaml:22]: I made the ParadeDB test use the same MinIO bucket, to simplify things, but I need both tests to use different paths, as the tests run in parallel.

Comment (philippemnoel) [reply charts/cluster/test/postgresql-minio-backup-restore/05-recovery_backup_cluster.yaml:22]: Sounds good

File: charts/cluster/README.md [hunk 8]
```diff
@@ -236,7 +239,10 @@ refer to  the [CloudNativePG Documentation](https://cloudnative-pg.io/documentat
 | recovery.s3.secretKey | string | `""` |  |
 | recovery.secret.create | bool | `true` | Whether to create a secret for the backup credentials |
 | recovery.secret.name | string | `""` | Name of the backup credentials secret |
-| type | string | `"postgresql"` | Type of the CNPG database. Available types: * `postgresql` * `postgis` |
+| type | string | `"postgresql"` | Type of the CNPG database. Available types: * `postgresql` * `postgis` * `timescaledb` |
```
Comment (philippemnoel) [inline charts/cluster/README.md:242]: This should add ParadeDB, no?

File: charts/cluster/test/paradedb-minio-backup-restore/02-paradedb_write.yaml (hunk 4 above)
Comment (itay-grudev) [reply charts/cluster/test/paradedb-minio-backup-restore/02-paradedb_write.yaml:1]: Isn't this creating a whole new table?
```sql
CALL paradedb.create_bm25_test_table( schema_name => 'public', table_name => 'mock_items' );
```

Although I am open to suggestions for a rename.

File: charts/cluster/test/paradedb-minio-backup-restore/03-paradedb_test.yaml (hunk 5 above)
Comment (itay-grudev) [reply charts/cluster/test/paradedb-minio-backup-restore/03-paradedb_test.yaml:26]: In my opinion, the purpose of this is not to verify that the functionality is working, but that it's simply there. Ensuring that we are using the correct Docker image with ParadeDB extensions. You should have a separate test suite that ensures the functionality.

But of course I can add all of the examples, but writing the bash check for each individual example will take time.

File: charts/cluster/README.md (hunk 8 above)
Comment (itay-grudev) [reply charts/cluster/README.md]: We should!

File: charts/cluster/values.yaml (hunk 7 above)
Comment (itay-grudev) [reply charts/cluster/values.yaml:22]: Can we do this in a separate PR?

Comment (itay-grudev) [reply charts/cluster/values.yaml:22]: Or can we implement my proposal where we only pin a minor version like: ` paradedb: "0.9"`

Comment (philippemnoel) [reply charts/cluster/values.yaml:22]: Yes, I can handle it in a separate PR. Leave it to me

File: charts/cluster/README.md.gotmpl [hunk 9]
```diff
@@ -69,6 +69,8 @@ Cluster Configuration
 Currently the chart supports two database types. These are configured via the `type` parameter. These are:
 * `postgresql` - A standard PostgreSQL database.
 * `postgis` - A PostgreSQL database with the PostGIS extension installed.
+* `timescaledb` - A PostgreSQL database with the TimescaleDB extension installed.
+* `paradedb` - A PostgreSQL database with the ParadeDB extension installed.
```
Comment (philippemnoel) [inline charts/cluster/README.md.gotmpl:73]:
```suggestion
* `paradedb` - A PostgreSQL database with the ParadeDB extensions (`pg_search` and `pg_analytics`) installed.
```

File: charts/cluster/test/paradedb-minio-backup-restore/06-post_backup_data_write.yaml [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+            valueFrom:
+              fieldRef:
+                fieldPath: metadata.namespace
+        image: alpine:3.19
+        command: ['sh', '-c']
+        args:
+         - |
+           apk --no-cache add postgresql-client
+           DB_URI=$(echo $DB_URI | sed "s|/\*|/|" )
+           psql "$DB_URI" -c "CREATE TABLE mybadtable (id serial PRIMARY KEY);"
```
Comment (philippemnoel) [inline charts/cluster/test/paradedb-minio-backup-restore/06-post_backup_data_write.yaml:27]: newline

File: charts/cluster/values.yaml (hunk 7 above)
Comment (philippemnoel) [reply charts/cluster/values.yaml:22]: #17

File: charts/cluster/test/paradedb-minio-backup-restore/02-paradedb_write.yaml [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+                schema_name => 'public',
+                table_name => 'mock_items',
+                key_field => 'id',
+                text_fields => paradedb.field('description', tokenizer => paradedb.tokenizer('en_stem')) ||
+                               paradedb.field('category'),
+                numeric_fields => paradedb.field('rating')
+              );
+            EOSQL
+
+
```
Comment (philippemnoel) [inline charts/cluster/test/paradedb-minio-backup-restore/02-paradedb_write.yaml:35]: extra newline

File: charts/cluster/test/paradedb-minio-backup-restore/02-paradedb_write.yaml (hunk 4 above)
Comment (philippemnoel) [reply charts/cluster/test/paradedb-minio-backup-restore/02-paradedb_write.yaml:1]: Sure, it does create a new table. Perhaps `bootstrap`, but either is fine.

Other question: I don't see anywhere the rest of the ParadeDB Quickstart be run. Could we add the other queries as a test as well so know that we can successfully run the full quickstart?

File: charts/cluster/test/paradedb-minio-backup-restore/03-paradedb_test.yaml (hunk 5 above)
Comment (philippemnoel) [reply charts/cluster/test/paradedb-minio-backup-restore/03-paradedb_test.yaml:26]: That's fair, we do have a test suite for it in the Dockerfile. This is good enough.

**Review by philippemnoel — APPROVED** [review]
lgtm. THANK YOU for doing this!

File: charts/cluster/test/paradedb-minio-backup-restore/02-paradedb_write.yaml (hunk 4 above)
Comment (philippemnoel) [reply charts/cluster/test/paradedb-minio-backup-restore/02-paradedb_write.yaml:1]: Write is fine

**Pull request merged**